File: PhoneInput/phoneinputlib/src/main/java/com/chinalwb/usphoneinput/injava/USPhoneInputEditText.java
package com.chinalwb.usphoneinput.injava;

import android.content.Context;
import android.text.*;
import android.util.AttributeSet;
import androidx.appcompat.widget.AppCompatEditText;
import io.michaelrocks.libphonenumber.android.NumberParseException;
import io.michaelrocks.libphonenumber.android.PhoneNumberUtil;
import io.michaelrocks.libphonenumber.android.Phonenumber;

/**
 * An extension on EditText. Accept digits only. Format the input as US phone number format
 * automatically. (XXX) XXX-XXXX.
 *
 * @author Wenbin
 */
public class USPhoneInputEditText extends AppCompatEditText {

    private static final String BRACKET = "(";
    private boolean ignoreChange = false;
    private boolean isBackspace = false;
    private String beforeChangeString = "";
    private int beforeChangeStart = -1;
    private int beforeChangeCount = -1;
    private PhoneNumberUtil util = PhoneNumberUtil.createInstance(getContext());

    public USPhoneInputEditText(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
        init();
    }

    public USPhoneInputEditText(Context context, AttributeSet attributeSet, int defStyleAttr) {
        super(context, attributeSet, defStyleAttr);
        init();
    }

    private void init() {
        this.setInputType(InputType.TYPE_CLASS_NUMBER);
        this.setFilters(new InputFilter[]{
                new InputFilter.LengthFilter(14)
        });
        setupListeners();
    }

    private void setupListeners() {
        this.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (after < count) {
                    isBackspace = true;
                    beforeChangeString = s.toString();
                    beforeChangeStart = start;
                    beforeChangeCount = count;
                } else {
                    isBackspace = false;
                    beforeChangeString = "";
                    beforeChangeStart = -1;
                    beforeChangeCount = -1;
                }
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int start, int before, int count) {
                // Do nothing
            }

            @Override
            public void afterTextChanged(Editable s) {
                // Handle delete
                if (isBackspace
                        && !TextUtils.isEmpty(beforeChangeString)
                        && beforeChangeStart > -1
                        && beforeChangeCount > -1) {
                    if (TextUtils.isEmpty(s)) {
                        return;
                    }
                    if (s.toString().equalsIgnoreCase(BRACKET)) {
                        setText("");
                        return;
                    }
                    String deletedContent = beforeChangeString.substring(
                            beforeChangeStart, beforeChangeStart + beforeChangeCount);
                    String deletedContentWithoutNumbers = deletedContent.replaceAll("[0-9]", "");

                    // There is no numbers being deleted
                    if (deletedContent.length() == deletedContentWithoutNumbers.length()) {
                        isBackspace = false;
                        setText(beforeChangeString);
                        endFocus();
                        return;
                    }

                    // There are some numbers are being deleted
                    String leftPart = beforeChangeString.substring(0, beforeChangeStart);
                    String rightPart = beforeChangeString.substring(beforeChangeStart + beforeChangeCount);
                    String finalString = leftPart + rightPart;
                    finalString = formatAsPhone(finalString);
                    if (TextUtils.isEmpty(finalString)) {
                        return;
                    }
                    setText(finalString);
                    endFocus();
                    return;
                }

                // Handle insert
                if (null == getText()) {
                    return;
                }
                String text = getText().toString();
                int textLength = text.length();
                if (text.endsWith("-") || text.endsWith(" ")) {
                    return;
                }
                try {
                    Phonenumber.PhoneNumber phoneNumber = util.parse(s.toString(), "US");
                    boolean isValid = util.isValidNumber(phoneNumber);
                    if (isValid) {
                        if (!ignoreChange) {
                            ignoreChange = true;
                            String formatResult = formatAsPhone(s.toString());
                            setText(formatResult);
                            endFocus();
                        }
                        return;
                    } else {
                        ignoreChange = false;
                    }
                } catch (NumberParseException e) {
                    e.printStackTrace();
                }

                ignoreChange = false;
                if (textLength == 1) {
                    if (!text.contains("(")) {
                        setText(new StringBuilder(text).insert(text.length() - 1, "(").toString());
                        setSelection(getText().length());
                    }
                } else if (textLength == 5) {
                    if (!text.contains(")")) {
                        setText(new StringBuilder(text).insert(text.length() - 1, ")").toString());
                        setSelection(getText().length());
                    }
                } else if (textLength == 6) {
                    setText(new StringBuilder(text).insert(text.length() - 1, " ").toString());
                    setSelection(getText().length());
                } else if (textLength == 10) {
                    if (!text.contains("-")) {
                        setText(new StringBuilder(text).insert(text.length() - 1, "-").toString());
                        setSelection(getText().length());
                    }
                } else {
                    String formatResult = formatAsPhone(s.toString());
                    if (null != formatResult && formatResult.equals(s.toString())) {
                        return;
                    }
                    setText(formatResult);
                    setSelection(getText().length());
                }

            }
        });
    }

    private void endFocus() {
        if (null != getText()) {
            setSelection(getText().length());
        }
    }

    private String formatAsPhone(String phoneNumberString) {
        phoneNumberString = phoneNumberString.replaceAll("[^\\d]", "");
        phoneNumberString = phoneNumberString.replaceFirst("(\\d{3})(\\d{3})(\\d+)", "($1) $2-$3");
        if (!phoneNumberString.contains(BRACKET)) {
            phoneNumberString = phoneNumberString.replaceFirst("(\\d{3})(\\d+)", "($1) $2");
            if (!phoneNumberString.contains(BRACKET)) {
                phoneNumberString = phoneNumberString.replaceFirst("(\\d+)", "($1");
            }
        }
        return phoneNumberString;
    }

    /**
     * Check if the current content in EditText is a valid US phone.
     *
     * @return true if it is a valid phone number; false otherwise
     */
    public boolean isValidUSPhone() {
        try {
            Phonenumber.PhoneNumber phoneNumber = util.parse(getText(), "US");
            return util.isValidNumber(phoneNumber);
        } catch (NumberParseException e) {
            e.printStackTrace();
            return false;
        }
    }
}

File: PhoneInput/phoneinputlib/src/main/java/com/chinalwb/usphoneinput/USPhoneInputEditText.kt
package com.chinalwb.usphoneinput

import android.content.Context
import android.text.*
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatEditText
import io.michaelrocks.libphonenumber.android.NumberParseException
import io.michaelrocks.libphonenumber.android.PhoneNumberUtil
import io.michaelrocks.libphonenumber.android.Phonenumber

class USPhoneInputEditText(context: Context, attributeSet: AttributeSet) : AppCompatEditText(context, attributeSet) {

    private var ignoreChange = false
    private var isBackspace = false
    private var beforeChangeString: String = ""
    private var beforeChangeStart = -1
    private var beforeChangeCount = -1
    private var util = PhoneNumberUtil.createInstance(context)

    init {
        this.inputType = InputType.TYPE_CLASS_NUMBER
        this.filters += InputFilter.LengthFilter(14)
        setupListeners()
    }


    private fun setupListeners() {
        this.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                if (after < count) {
                    isBackspace = true
                    beforeChangeString = s.toString()
                    beforeChangeStart = start
                    beforeChangeCount = count
                } else {
                    isBackspace = false
                    beforeChangeString = ""
                    beforeChangeStart = -1
                    beforeChangeCount = -1
                }
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                if (isBackspace && !TextUtils.isEmpty(beforeChangeString) && beforeChangeStart > -1 && beforeChangeCount > -1) {
                    if (TextUtils.isEmpty(s)) {
                        return
                    }
                    if (s.toString() == "(") {
                        setText("")
                        return
                    }
                    val deletedContent = beforeChangeString.substring(
                        beforeChangeStart,
                        beforeChangeStart + beforeChangeCount
                    )
                    val deletedContentWithoutNumbers =
                        deletedContent.replace("[0-9]".toRegex(), "")

                    // There is no numbers being deleted
                    if (deletedContent.length == deletedContentWithoutNumbers.length) {
                        isBackspace = false
                        setText(beforeChangeString)
                        setSelection(text!!.length)
                        return
                    }

                    // There are some numbers are being deleted
                    val leftPart = beforeChangeString.substring(0, beforeChangeStart)
                    val rightPart =
                        beforeChangeString.substring(beforeChangeStart + beforeChangeCount)
                    var finalString = leftPart + rightPart
                    finalString = formatAsPhone(finalString)
                    if (TextUtils.isEmpty(finalString)) {
                        return
                    }
                    setText(finalString)
                    setSelection(text!!.length)
                    return
                }

                val textLength = text!!.length
                if (text!!.endsWith("-") || text!!.endsWith(" "))
                    return
                try {
                    val phoneNumber = util.parse(s.toString(), "US")
                    val isValid = util.isValidNumber(phoneNumber)
                    if (isValid) {
                        if (!ignoreChange) {
                            ignoreChange = true
                            val formatResult = formatAsPhone(s.toString())
                            setText(formatResult)
                            setSelection(formatResult.length)
                        }
                        return
                    } else {
                        ignoreChange = false
                    }
                } catch (e: NumberParseException) {
                    e.printStackTrace()
                }

                ignoreChange = false
                if (textLength == 1) {
                    if (!text!!.contains("(")) {
                        setText(
                            StringBuilder(text!!).insert(
                                text!!.length - 1,
                                "("
                            ).toString()
                        )
                        setSelection(text!!.length)
                    }
                } else if (textLength == 5) {
                    if (!text!!.contains(")")) {
                        setText(
                            StringBuilder(text!!).insert(
                                text!!.length - 1,
                                ")"
                            ).toString()
                        )
                        setSelection(text!!.length)
                    }
                } else if (textLength == 6) {
                    setText(StringBuilder(text!!).insert(text!!.length - 1, " ").toString())
                    setSelection(text!!.length)
                } else if (textLength == 10) {
                    if (!text!!.contains("-")) {
                        setText(
                            StringBuilder(text!!).insert(
                                text!!.length - 1,
                                "-"
                            ).toString()
                        )
                        setSelection(text!!.length)
                    }
                } else if (textLength == 15) {
                    if (text!!.contains("-")) {
                        setText(
                            StringBuilder(text!!).insert(
                                text!!.length - 1,
                                "-"
                            ).toString()
                        )
                        setSelection(text!!.length)
                    }
                } else if (textLength == 18) {
                    if (text!!.contains("-")) {
                        setText(
                            StringBuilder(text!!).insert(
                                text!!.length - 1,
                                "-"
                            ).toString()
                        )
                        setSelection(text!!.length)
                    }
                } else {
                    val formatResult = formatAsPhone(s.toString())
                    if (formatResult == s.toString()) {
                        return
                    }
                    setText(formatResult)
                    setSelection(formatResult.length)
                }
            }
        })
    }


    private fun formatAsPhone(inputString: String): String {
        var phoneNumberString = inputString
        phoneNumberString = phoneNumberString.replace("[^\\d]".toRegex(), "")
        phoneNumberString =
            phoneNumberString.replaceFirst("(\\d{3})(\\d{3})(\\d+)".toRegex(), "($1) $2-$3")
        val bracket = "("
        if (!phoneNumberString.contains(bracket)) {
            phoneNumberString =
                phoneNumberString.replaceFirst("(\\d{3})(\\d+)".toRegex(), "($1) $2")
            if (!phoneNumberString.contains(bracket)) {
                phoneNumberString = phoneNumberString.replaceFirst("(\\d+)".toRegex(), "($1")
            }
        }
        return phoneNumberString
    }

    fun isValidUSPhone() : Boolean {
        return try {
            val phoneNumber = util.parse(text!!, "US")
            util.isValidNumber(phoneNumber)
        } catch (e: NumberParseException) {
            e.printStackTrace()
            false
        }
    }
}